applets/wncklet/window-list.c

```diff
@@ -255,50 +257,71 @@ static GdkPixbuf *preview_window_thumbnail (WnckWindow *wnck_window, TasklistDat
 	width = gdk_window_get_width (window) * scale;
 	height = gdk_window_get_height (window) * scale;
 
-	/* Generate window screenshot for preview */
-	screenshot = gdk_pixbuf_get_from_window (window, 0, 0, width / scale, height / scale);
+	/* Get reference to GdkWindow surface */
+	cairo_t *win_cr = gdk_cairo_create (window);
+	cairo_surface_t *win_surface = cairo_get_target (win_cr);
+	/* Flush to ensure all writing to the image was done */
+	cairo_surface_flush (win_surface);
+	cairo_destroy (win_cr);
+
+	/* Create screenshot surface with the GdkWindow as its source */
+	screenshot = cairo_image_surface_create (CAIRO_FORMAT_ARGB32, width, height);
+	cairo_surface_set_device_scale (screenshot, scale, scale);
+	cr = cairo_create (screenshot);
+	cairo_set_source_surface (cr, win_surface, 0, 0);
+	cairo_paint (cr);
+
+	/* Mark the image dirty so Cairo clears its caches */
+	cairo_surface_mark_dirty (win_surface);
+
+	cairo_destroy (cr);
 	g_object_unref (window);
 
 	if (screenshot == NULL)
 		return NULL;
 
-	/* Determine whether the contents of the screenshot are empty */
-	if (gdk_pixbuf_get_byte_length (screenshot) == 0)
-	{
-		g_object_unref (screenshot);
-		return NULL;
-	}
-
 	/* Scale to configured size while maintaining aspect ratio */
 	if (width > height)
 	{
-		ratio = (double) height / (double) width;
-		width = MIN(width, tasklist->thumbnail_size);
-		height = width * ratio;
+		int max_size = MIN(width, tasklist->thumbnail_size);
+		ratio = (double) width / (double) max_size;
+		width = max_size;
+		height = (int) ((double) height / ratio);
```

Comment:
The ratio is usually within the range 0..1, which can be achieved by exchanging the numerator and denominator and multiplying the smaller of the two by the ratio later. In addition, the computational cost of a floating-point division is usually higher than a multiplication.

Reply:
```diff
diff --git a/applets/wncklet/window-list.c b/applets/wncklet/window-list.c
index bb910f20..fdfa72a8 100644
--- a/applets/wncklet/window-list.c
+++ b/applets/wncklet/window-list.c
@@ -283,17 +283,17 @@ static cairo_surface_t *preview_window_thumbnail (WnckWindow *wnck_window, Taskl
        /* Scale to configured size while maintaining aspect ratio */
        if (width > height)
        {
-               int max_size = MIN(width, tasklist->thumbnail_size);
-               ratio = (double) width / (double) max_size;
+               int max_size = MIN (width, tasklist->thumbnail_size);
+               ratio = (double) max_size / (double) width;
                width = max_size;
-               height = (int) ((double) height / ratio);
+               height = (int) ((double) height * ratio);
        }
        else
        {
-               int max_size = MIN(height, tasklist->thumbnail_size);
-               ratio = (double) height / (double) max_size;
+               int max_size = MIN (height, tasklist->thumbnail_size);
+               ratio = (double) max_size / (double) height;
                height = max_size;
-               width = (int) ((double) width / ratio);
+               width = (int) ((double) width * ratio);
        }
 
        thumbnail = cairo_surface_create_similar (screenshot,
@@ -301,7 +301,7 @@ static cairo_surface_t *preview_window_thumbnail (WnckWindow *wnck_window, Taskl
                                                  width, height);
 
        cr = cairo_create (thumbnail);
-       cairo_scale (cr, scale/ratio, scale/ratio);
+       cairo_scale (cr, scale * ratio, scale * ratio);
        cairo_set_source_surface (cr, screenshot, 0, 0);
        cairo_paint (cr);
 
```

Reply:
Thanks, good catch :+1: